alws/crud.py

```diff
@@ -867,3 +874,223 @@ async def commit_release(db: Session, release_id: int) -> (models.Release, str):
+                repos_mapping[repo_key] = models.Repository(**repo_item_dict)
+            else:
+                repo = repos_mapping[repo_key]
+                for field, value in repo_item_dict.items():
+                    setattr(repo, field, value)
+
+        db.add_all(repos_mapping.values())
+        await db.commit()
+
+    for repo in repos_mapping.values():
+        await db.refresh(repo)
+
+    return list(repos_mapping.values())
+
+
+async def create_repository(
+        db: Session, payload: repository_schema.RepositoryCreate,
+) -> models.Repository:
+    query = select(models.Repository).where(
+        models.Repository.name == payload.name,
+        models.Repository.arch == payload.arch,
+        models.Repository.type == payload.type,
+        models.Repository.debug == payload.debug,
+    )
+    async with db.begin():
+        result = await db.execute(query)
+        if result.scalars().first():
+            raise ValueError('Repository already exists')
+
+        repository = models.Repository(**payload.dict())
+        db.add(repository)
+        await db.commit()
+    await db.refresh(repository)
+    return repository
+
+
+async def search_repository(
+        db: Session, payload: repository_schema.RepositorySearch
+) -> models.Repository:
+    query = select(models.Repository)
+    for key, value in payload.dict().items():
+        if key == 'name':
+            query = query.where(models.Repository.name == value)
+        elif key == 'arch':
+            query = query.where(models.Repository.arch == value)
+        elif key == 'type':
+            query = query.where(models.Repository.type == value)
+        elif key == 'debug':
+            query = query.where(models.Repository.debug == value)
+    async with db.begin():
+        result = await db.execute(query)
+        return result.scalars().first()
+
+
+async def update_repository(
+        db: Session, repository_id: int,
+        payload: repository_schema.RepositoryUpdate
+) -> models.Repository:
+    async with db.begin():
+        repository = await db.execute(select(
+            models.Repository).get(repository_id))
+        for field, value in payload.dict():
+            setattr(repository, field, value)
+        db.add(repository)
+        await db.commit()
+    await db.refresh(repository)
+    return repository
+
+
+async def delete_repository(db: Session, repository_id: int):
+    async with db.begin():
+        await db.execute(delete(models.Repository).where(
+            models.Repository.id == repository_id))
+        await db.commit()
+
+
+async def add_to_platform(db: Session, platform_id: int,
+                          repository_ids: typing.List[int]) -> models.Platform:
+    platform_result = await db.execute(select(models.Platform).where(
+        models.Platform.id == platform_id).options(
+        selectinload(models.Platform.repos)).with_for_update())
+    platform = platform_result.scalars().first()
```

Comment:
Platform could be not found in general case

Reply:
Yeah, you're right, it may be, will fix